### ISSUE TITLE
Events Access Control

### ISSUE THREAD

**Issue by klizhentas** [gravitational/teleport#5430]
### Feature Request

Update events system to support the following use-cases:

* Members of the teams should be able to see events from their own team, but not others (including session recordings and playbacks).

* Create a role for team auditors, who will be able to see the events from some teams, but not access the infrastructure.

* Provide efficient pagination so people can export the data (current API has hardcoded limit per range period that makes data export a problem) #5435 

## Backends

* For on-premises, customers are OK to use Elastic Search as a backend.
* We have to support existing backends, DynamoDB and local disk backend.

## Namespaces

I would like to finish implementation of namespaces to fix this problem. Events, nodes and every single resource already support namespaces internally. The default namespace is used at all times at the moment.

Namespaces design should be similar to Kubernetes: 

* Nodes, Kubernetes clusters, DBs and Apps could be placed into namespace.
* Default namespaces is used by default as today
* Users can select a namespace and `tsh profile` could be updated to select the namespace.
* Events and session recordings generated by users accessing the resources should be placed in the namespace
* Roles are updated to support access to namespaces:

```yaml
kind: role
name: team-access
spec:
   allow:
      namespaces: ['{{traits.teams}}']
```

* CLI flags 
  `tsh ls --n default`
   `tsh set-context --set-namespace=default --set-cluster=root`
  `tctl get nodes -n default`

* User interface. A new namespace selector should be present in the UI next to the cluster selector.

**Comment by klizhentas** [comment]
## DynamoDB

The biggest headache in this model is dynamodb. Our Search API query looks like this:

```
EventNamespace = :eventNamespace AND createdAt BETWEEN :start and :end
```

Currently event namespace is always `default`. This query hits the global secondary index in DynamoDB and is reasonably efficient.

In DynamoDB the primary, partition key has to be queried with [eq condition](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.html#Query.KeyConditionExpressions)

> You must specify the partition key name and value as an equality condition.

The only way to support filtering by team in DynamoDB is to create a global secondary index. For example, we can let teams define `labels` in the events. We can then create a global secondary index that will have a global secondary index for each label value and allow queries like:

```
#GlobalSecondaryIndex for label team
EventLabelKey = :eventLabelValue AND createdAt BETWEEN :start and :end
```

We would have to put the limit on the amount of different label keys for events, because otherwise there will be too many global secondary indexes on the events.

**Comment by klizhentas** [comment]
The alternative is to go Kubernetes style and bring back namespaces concept. Namespaces provide better isolation, but this could require a large refactor.

**Comment by russjones** [comment]
@klizhentas While namespaces exist, since they have been unused since inception, I don't know how much isolation they provide. They may be in a more unfinished state than we expect since their functionality has never been exercised.

The other thing is namespaces force the user to think about them, ideally we wouldn't do that to users.

We could add tags to events then in our resource definition we could use the `where` clause to specify that this user can fetch `events` but only events with tags matching some value that comes from claims. Admittedly this would work for ElasticSearch, probably not DynamoDB and disk.

**Comment by klizhentas** [comment]
Namespaces provide better concept for team isolation and security. Kubernetes is a great example. Once you place a resource in a namespace, it is really not visible by default and all actions on it require namespace to be defined. It is true that namespaces force people to think about them. However, imagine the most common use-case - a team gets a namespace in teleport. Most of the time that's what they will use and they won't need to think about the other namespaces.

TLDR: Kubernetes has proven that namespaces are a good idea for isolation purposes. Teleport on purpose uses similar concepts like labels, so bringing in namespaces is a good choice.

**Comment by klizhentas** [comment]
Here is a cheaper version of it:

* Have every event associated with a user record user and user's traits
* Have every node associated with a user record a node and node's traits

Support a role:

```yaml
kind: role
version: v3
metadata:
  name: session-viewer
spec:

    # list of allow-rules. see below for more information.
    rules:
    - resources: [session]
      verbs: [list, read]
      where: 'contains(event.spec.user.labels, user.spec.traits["group"])'
   # allow to see events that are initiated by user's group
    - resources: [event]
      verbs: [list, read]
      where: 'contains(event.spec.user.labels, user.spec.traits["group"])'
```

**Comment by russjones** [comment]
@klizhentas

1. This would apply to session recordings, do we want to do any restrictions on joining a live session? It's a different resource `ssh_session`.

2. I don't think all events in the system are emitted with a user identity, if I recall correctly some are just cluster state change information. The only way to see those would be if you don't have a `where` clause specified which means wildcard right?

```yaml
kind: role
version: v3
metadata:
  name: admin
spec:
    rules:
    - resources: [event]
      verbs: [list, read]
```

3. Do we still use `GetSessionEvents` and `GetSessionChunk`? I think those are the only ones that use the `session` resource. If those were only used pre-streaming, we can just use `event`.

**Comment by klizhentas** [comment]
@stevenGravy @russjones 

MVP:

Have modes:

* Users can only view their only session recordings
* Users can't join other users sessions and view active sessions

Post MVP:

* Have mode when team members can view and join sessions based on the traits and labels

**Comment by klizhentas** [comment]
* See no problem in adding another index to DynamoDB to solve the Access Control server side problem.

**Comment by russjones** [comment]
RFD for Events Access Control: https://github.com/gravitational/teleport/pull/8084

**Comment by russjones** [comment]
Remaining backend changes for active sessions @codingllama.

https://github.com/gravitational/teleport/pull/8588
https://github.com/gravitational/teleport/pull/8568

Remaining frontend changes for recorded and active sessions @atburke.

https://github.com/gravitational/webapps/issues/406

**Comment by codingllama** [comment]
Active sessions RFD merged as https://github.com/gravitational/teleport/pull/8962 - I'm trying out the server-side changes, they may need a couple tweaks.

**Comment by codingllama** [comment]
Active sessions implementation landed on master yesterday, RFDs are up-to-date, so all the backend work required here is done. I'm working on backporting changes to v8, but that should be simple (and doesn't block anyone).

@rudream, something I noticed in the Web UI is that when a user tries to join a forbidden active session (say, because of a new "where" condition), it errors saying "The session is no longer active" instead of giving an "access denied" error. Just giving you a heads up in case you want to tweak that.

**Comment by russjones** [comment]
I'm going to re-open this but remove it from the Queue. We implemented basic support for events access control (can see your own or all sessions) but we didn't implement team based access controls where all members of a particular team can see each other sessions.

**Comment by zmb3** [comment]
We now support advanced RBAC on session recordings, so we're only lacking more control over audit events.

We have #10405 for tracking audit events, so I'm closing this issue in favor of that one.
